File: q1-two-sum/q1-hash-method.cpp
// LeetCode 1.
// https://leetcode.com/problems/two-sum/
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        vector<int> ans;
        unordered_map<int, int> mapping;
        // store {vlue: position}

        for (int i = 0; i < nums.size(); i++)
        {
            int corValue = target - nums[i];

            // if the correndence value exists in the map
            if (mapping.count(corValue))
            {
                ans.push_back(mapping[corValue]);
                ans.push_back(i);
                break;
            }
            mapping[nums[i]] = i;
        }
        return ans;
    }
};

// time: O(n)
// space: O(n)

int main()
{
    vector<int> case1{2, 7, 11, 15};

    Solution sol;
    vector<int> ans = sol.twoSum(case1, 9);

    for (int num : ans)
    {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

File: ztm-ds-algo/1-arrays/q1-two-sum/q1-hash-method.js
// LeetCode 1.
// https://leetcode.com/problems/two-sum/

const twoSum = function(nums, target) {
    const ans = [];
    const mapping = new Map() // store {value: position}

    for (let i = 0; i< nums.length; i++) {
        let com = target - nums[i];

        if (mapping.has(nums[i])) {
            return [mapping.get(nums[i]), i];
        }

        mapping.set(com, i);
    }
}

// time: O(n)
// space: O(n)